core-web/src/main/java/org/silverpeas/core/web/mvc/controller/AbstractAdminComponentSessionController.java

```diff
@@ -0,0 +1,208 @@
+   * <p>
+   *   Each implementation can precise or change this default implementation.
+   * </p>
+   * @return true if access granted, false otherwise.
+   */
+  public boolean isAccessGranted() {
+    return isUserAdmin();
+  }
+
+  /**
+   * This method is invoked at each administration service access.
+   * @throws ForbiddenRuntimeException in case the user has forbidden access.
+   */
+  public void checkAccessGranted() {
+    if (!isAccessGranted()) {
+      throwForbiddenError();
+    }
+  }
+
+  /**
+   * Indicates if the user can access administration about a space or a component instance.
+   * <p>
+   *   Whatever the parameters, user having admin access right on its account is always access
+   *   granted.
+   * </p>
+   * <p>
+   *   If both space id and instance id are given, they are both verified.
+   * </p>
+   * <p>
+   *   If neither space if neither instance id are given, it is verified that current user is a
+   *   space manager.
+   * </p>
+   * @param spaceId the optional identifier of a space.
+   * @param instanceId the optional identifier of a component instance.
+   * @param readOnly true if the operation is read only, false otherwise.
+   * @return true if access granted, false otherwise.
+   */
+  public boolean isAccessGranted(final String spaceId, final String instanceId,
+      final boolean readOnly)
+      throws ForbiddenRuntimeException {
+    boolean accessGranted = getUserDetail().isAccessAdmin();
+    if (!accessGranted) {
+      final String[] userManageableSpaceIds = getUserManageableSpaceIds();
+      accessGranted = userManageableSpaceIds.length > 0 &&
+          (readOnly || isDefined(spaceId) || isDefined(instanceId));
+      if (accessGranted) {
+        final Set<String> userManageableSpaceIdSet = of(userManageableSpaceIds).collect(toSet());
+        if (isDefined(instanceId)) {
+          accessGranted = getOrganisationController().getComponentInstance(instanceId)
+              .map(SilverpeasComponentInstance::getSpaceId)
+              .map(userManageableSpaceIdSet::contains)
+              .orElse(false);
+        }
+        if (accessGranted && isDefined(spaceId)) {
+          accessGranted = ofNullable(getOrganisationController().getSpaceInstById(spaceId))
+              .map(SpaceInst::getId)
+              .map(s -> userManageableSpaceIdSet.contains(s) ||
+                  (readOnly && of(userManageableSpaceIds)
+                      .map(getOrganisationController()::getSpaceInstById)
+                      .filter(Objects::nonNull)
+                      .map(SpaceInst::getDomainFatherId)
+                      .filter(StringUtil::isDefined)
+                      .filter(not(userManageableSpaceIdSet::contains))
+                      .flatMap(m -> getOrganisationController().getPathToSpace(m).stream())
+                      .map(SpaceInstLight::getId)
+                      .anyMatch(s::equals)))
+              .orElse(false);
+        }
```

Comment:
Mmm... I require some explanations here. What I understand from the above code:
access is granted if:
- the space is among the spaces the user is able to manage
- or the access is read-only (not within a side-effect operation) and the parent of the space isn't among the spaces the user is able to manage but there is at least one parent to this parent that is manageable by the user.
What I don't understand is why the parent of the current space to check must not be among the the manageable spaces.

Reply:
`userManageableSpaceIdSet` contains the identifier of spaces the user can manage.
For example, if the user is set explicitly on `Space A` `userManageableSpaceIdSet` contains `Space A` identifier and those of `Space A` children.
In read only case, parent space of `Space A` must be accessed by the user without giving him writing rights. So parent spaces of spaces contained into `userManageableSpaceIdSet` have to be verified.
Each space path of space indentifiers included into the set MUST be verified. For performances, If the parent identifier of space identifier of the set is already itself in the set, there is no need to check the path as it has been verified on the previous if clause.